networkx/algorithms/dag.py

```diff
@@ -355,6 +356,50 @@ def transitive_closure(G):
 
 
 @not_implemented_for('undirected')
+def transitive_reduction(G):
+    """ Returns transitive reduction of a directed graph
+
+    The transitive reduction of G = (V,E) is a graph G- = (V,E-) such that
+    for all v,w in V there is an edge (v,w) in E- if and only if (v,w) is
+    in E and there is no path from v to w in G with length greater than 1.
+
+    Parameters
+    ----------
+    G : NetworkX DiGraph
+        Graph
+
+    Returns
+    -------
+    TR : NetworkX DiGraph
+        Graph
+
+    Raises
+    ------
+    NetworkXError
+        If G is not a directed acyclic graph (DAG) transitive reduction is
+        not uniquely defined and a NetworkXError exception is raised.
+
+    References 
+    ----------
+    https://en.wikipedia.org/wiki/Transitive_reduction
+
+    """   
+    if not is_directed_acyclic_graph(G):
+        raise nx.NetworkXError(
+            "Transitive reduction only uniquely defined on directed acyclic graphs.") 
+    TR = nx.DiGraph()
+    TR.add_nodes_from(G.nodes())
+    for u in G:
+        u_edges = set(G[u])
+        for v in G[u]:
+            for w in nx.dfs_preorder_nodes(G, source=v):
```

Comment:
I think you could avoid the `w != v` check below by doing something like `itertools.islice(nx.dfs_preorder_nodes(G, source=v), 1, None):`, since `v` is guaranteed to be the first element yielded by the iterator.